src/v/kafka/server/handlers/delete_records.cc

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/handlers/delete_records.h"
+
+#include "cluster/metadata_cache.h"
+#include "cluster/partition_manager.h"
+#include "cluster/shard_table.h"
+#include "model/fundamental.h"
+
+#include <vector>
+
+namespace kafka {
+
+/// Returned in responses where kafka::error_code is anything else then a value
+/// of error_code::none
+static constexpr auto invalid_low_watermark = model::offset(-1);
+
+/// Compare against user provided value of truncation offset, this value will
+/// indicate to truncate at  the current partition high watermark
+static constexpr auto current_high_watermark = model::offset(-1);
+
+static std::vector<delete_records_partition_result>
+make_partition_errors(const delete_records_topic& t, error_code ec) {
+    std::vector<delete_records_partition_result> r;
+    for (const auto& p : t.partitions) {
+        r.push_back(delete_records_partition_result{
+          .partition_index = p.partition_index,
+          .low_watermark = invalid_low_watermark,
+          .error_code = ec});
+    }
+    return r;
+}
+
+/// Performs validation of topics, any failures will result in a list of
+/// partitions that all contain the identical error codes
+static std::vector<delete_records_partition_result>
+validate_at_topic_level(request_context& ctx, const delete_records_topic& t) {
+    const auto is_authorized = [&ctx](const delete_records_topic& t) {
+        return ctx.authorized(security::acl_operation::remove, t.name);
+    };
+    const auto is_deletable = [&ctx](const delete_records_topic& t) {
+        const auto cfg = ctx.metadata_cache().get_topic_cfg(
+          model::topic_namespace_view(model::kafka_namespace, t.name));
+        if (!cfg || !cfg->properties.cleanup_policy_bitflags) {
```

Comment:
[is_collectable](https://github.com/redpanda-data/redpanda/blob/dev/src/v/storage/ntp_config.h#L134) also checks for the cleanup_policy_bitflags, but handles things differently (returns that it is deletable if there is none set). I'm not sure, but do we need to copy that logic? I see this causing trouble in fixture tests, but without digging further I'm not sure it's a problem outside of tests

Reply:
Yeah i would saw we should operate the same way other consumers of that API would react